kbfstool/md_dump.go

```diff
@@ -0,0 +1,323 @@
+	if parts == nil {
+		return libkbfs.ImmutableRootMetadata{},
+			fmt.Errorf("Could not parse %q", input)
+	}
+
+	tlfPartStr := parts[1]
+	branchPartStr := parts[2]
+	revisionPartStr := parts[3]
+
+	tlfID, err := getTlfID(ctx, config, tlfPartStr)
+	if err != nil {
+		return libkbfs.ImmutableRootMetadata{}, err
+	}
+
+	branchID, err := getBranchID(ctx, config, tlfID, branchPartStr)
+	if err != nil {
+		return libkbfs.ImmutableRootMetadata{}, err
+	}
+
+	rev, err := getRevision(ctx, config, tlfID, branchID, revisionPartStr)
+	if err != nil {
+		return libkbfs.ImmutableRootMetadata{}, err
+	}
+
+	return mdGet(ctx, config, tlfID, branchID, rev)
+}
+
+func getUserString(
+	ctx context.Context, config libkbfs.Config, uid keybase1.UID) string {
+	username, _, err := config.KeybaseService().Resolve(
+		ctx, fmt.Sprintf("uid:%s", uid))
+	if err != nil {
+		printError("md dump", err)
+		return uid.String()
+	}
+	return fmt.Sprintf("%s (uid:%s)", username, uid)
+}
+
+func mdDumpOne(ctx context.Context, config libkbfs.Config,
+	rmd libkbfs.ImmutableRootMetadata) error {
+	mdID, err := config.Crypto().MakeMdID(&rmd.BareRootMetadata)
+	if err != nil {
+		return err
+	}
+
+	fmt.Printf("MD ID: %s\n", mdID)
+
+	buf, err := config.Codec().Encode(&rmd.BareRootMetadata)
+	if err != nil {
+		return err
+	}
+
+	fmt.Printf("MD size: %d bytes\n\n", len(buf))
+
+	fmt.Print("Reader/writer metadata\n")
+	fmt.Print("----------------------\n")
+	fmt.Printf("Last modifying user: %s\n",
+		getUserString(ctx, config, rmd.LastModifyingUser))
+	// TODO: Print flags.
+	fmt.Printf("Revision: %s\n", rmd.Revision)
+	fmt.Printf("Prev MD ID: %s\n", rmd.PrevRoot)
+	// TODO: Print RKeys, unresolved readers, conflict info,
+	// finalized info, and unknown fields.
+	fmt.Print("\n")
+
+	fmt.Print("Writer metadata\n")
+	fmt.Print("---------------\n")
+	fmt.Printf("Last modifying writer: %s\n",
+		getUserString(ctx, config, rmd.LastModifyingWriter))
+	// TODO: Print Writers/WKeys and unresolved writers.
+	fmt.Printf("TLF ID: %s\n", rmd.ID)
+	fmt.Printf("Branch ID: %s\n", rmd.BID)
+	// TODO: Print writer flags.
+	fmt.Printf("Disk usage: %d\n", rmd.DiskUsage)
+	fmt.Printf("Bytes in new blocks: %d\n", rmd.RefBytes)
+	fmt.Printf("Bytes in unreferenced blocks: %d\n", rmd.UnrefBytes)
+	// TODO: Print unknown fields.
+	fmt.Print("\n")
+
+	fmt.Print("Private metadata\n")
+	fmt.Print("----------------\n")
+	fmt.Printf("Serialized size: %d bytes\n", len(rmd.SerializedPrivateMetadata))
+
+	data := rmd.Data()
+	// TODO: Clean up output.
+	fmt.Printf("Dir: %s\n", data.Dir)
+	fmt.Print("TLF private key: {32 bytes}\n")
+	if data.ChangesBlockInfo() != (libkbfs.BlockInfo{}) {
+		fmt.Printf("Block changes block: %v\n", data.ChangesBlockInfo())
+	}
+	for i, op := range data.Changes.Ops {
+		fmt.Printf("Op[%d]: %v\n", i, op)
+	}
+	// TODO: Print unknown fields.
+
+	return nil
+}
+
+const mdDumpUsageStr = `Usage:
+  kbfstool md dump input [inputs...]
+
+Each input must be in the following format:
+
+  TlfPart
```

Comment:
I don't think you need the "Part" suffixes here either, it makes it more confusing (I keep thinking it means "partial" for some reason).